### PR TITLE
Improved workarounds for win32

### PR DESCRIPTION
**Pull request by mekwall**
### Summary
This PR improves on the existing override for `fs.rename` and also adds a similar but blocking override for `fs.renameSync` (obsoleting #23). The overrides tries to normalize the behavior of `fs.rename` and `fs.renameSync` between platforms. It does this by ensuring that the file was actually moved before resolving within a 60 second time frame.

### Background
`fs.rename` and `fs.renameSync` uses [MoveFileEx function](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365240(v=vs.85).aspx) on Windows. It is not an atomic operation and honors the Windows sharing modes, meaning that whenever a file or parent directory is locked (in use) the rename might fail with _EACCS_ or _EPERM_ errors depending on the sharing mode set on the file and/or directory.

This differs from how the api call used by Linux and OSX works where the rename operation is atomic and will go through no matter if the file is locked or not.

### Disabling anti-virus is NOT a viable solution
Most active anti-virus file scanners will lock the file and/or directory during scan, so if you are trying to rename it during this time you'll get an EPERM/EACCS error thrown in your face. Just to be clear: Disabling AV is NOT a viable solution and should never be accepted as such.

### Issues possibly affected by this
- https://github.com/isaacs/node-graceful-fs/issues/104
- https://github.com/isaacs/node-graceful-fs/issues/118
- https://github.com/npm/npm/issues/10826
- https://github.com/npm/npm/issues/16845
- https://github.com/npm/npm/issues/17321
- https://github.com/npm/write-file-atomic/issues/28
- https://github.com/atom/atom/issues/14913
- https://github.com/facebook/jest/issues/4444
- https://github.com/chrmarti/testissues/issues/1290
- https://github.com/typings/typings/issues/790
- https://github.com/yahoo/pure/issues/647
- https://github.com/arboleya/electrify/issues/75
- https://github.com/ember-cli/ember-cli/issues/6765
- https://github.com/Pana/nrm/issues/28

### References
- [node.js source on fs.rename](https://github.com/nodejs/node/blob/8485a7c0b71fd1dd36947a6dd02b75387805d9b7/deps/uv/src/win/fs.c#L1289)
- [Windows File API MoveFileEx function](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365240(v=vs.85).aspx)
- [rename(2) Linux man page](http://man7.org/linux/man-pages/man2/rename.2.html)

### CODE REVIEW

**Comment by mekwall** [comment]
For some reason I had to add unlink for when the source and destination both exist at the same time. But afaik `fs.rename` use the [MoveFile function](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365239(v=vs.85).aspx) on Windows and that should not copy the source to the destination. The source should just have it's path/name changed so the source and destination should never be able to exist both at once, correct me if I'm wrong.

File: polyfills.js [hunk 1]
```diff
@@ -99,11 +99,19 @@ function patch (fs) {
...
-                fs$rename(from, to, CB);
+            fs.stat(to, function (toStater, toSt) {
+              if (toStater && toStater.code === "ENOENT")
+                fs$rename(from, to, CB)
               else
-                cb(er)
+                fs.stat(from, function (fromStater, fromSt) {
+                  if (!fromSt && toSt) {
+                    if (cb) cb(toStater ? er : null)
+                  } else if (fromSt && toSt && fromSt.size === toSt.size) {
```
Comment (mekwall) [inline polyfills.js]: Not really fond about this, so could use some feedback/input on how it can be solved differently.

File: polyfills.js [hunk 2]
```diff
@@ -94,25 +94,58 @@ function patch (fs) {
...
+            while (waitUntil > Date.now()){}
+            tryRename()
+          } else if (backoff > 0 && e.code === "ENOENT") {
+            // The source does no longer exist because it was moved during one of the retries
+            // so we can safely ignore this exception
+          } else {
+            throw e
+          }
+          // Wait until destination exists and source no longer exists or that we've reached the backoff limit
+          while ((fs.existsSync(from) || !fs.existsSync(to)) && Date.now() < backoffUntil) {}
```
Comment (mekwall) [inline polyfills.js]: Possibly the worst hack ever, but it seems to be the only real way to solve inconsistent behavior on Windows where the destination and source can exist at the same time.

**Comment by conceptualspace** [comment]
testing this out on a windows machine lead to crash:

"error": "TypeError: Cannot read property 'size' of undefined", "errorStackTrace": "TypeError: Cannot read property 'size' of undefined\n    at node_modules\\graceful-fs\\polyfills.js:112:45\n    at node_modules\\graceful-fs\\polyfills.js:331:29\n    at FSReqWrap.oncomplete (fs.js:114:15)"

Node v7.9.0
Windows Server 2008

**Comment by mekwall** [comment]
@conceptualspace I've added some additional sanity checks that should solve your crash. Thanks for testing and reporting the issue!

**Comment by conceptualspace** [comment]
noticed a regression. see https://github.com/isaacs/node-graceful-fs/issues/98 for discussion (PR https://github.com/isaacs/node-graceful-fs/pull/100)

sometimes EPERM is expected -- when destination directory already exists windows will not overwrite, so in this case we do not want to retry but should fail immediately

**Comment by mekwall** [comment]
I disagree. `fs.rename` is supposed to replace an existing target and will do so as long as it does not run into any EACCS/EPERM issues, and this only happen on Windows, since rename on Linux/OSX doesn't care if the existing file is locked or not.

If it was expected to fail due to target already existing, it would throw `Error: EEXIST: file already exists`, which it doesn't.

**Comment by conceptualspace** [comment]
that would be fine if node actually reported an EEXIST error, but it does not on windows -- we just get EACCS/EPERM if the destination already exists

**Comment by conceptualspace** [comment]
```
> fs.mkdirSync('test1');
> fs.mkdirSync('test2');
> fs.rename('test2', 'test1', function(err) {console.log(err);});
Error: EPERM: operation not permitted, rename 'D:\Dev\konote\temp0' -> 'D:\Dev\konote\temp'
```
win 10x64
node v8.4.0

**Comment by mekwall** [comment]
This only happens for directories and if you do the same on Linux/OSX you'll see it works just fine since the native rename in Linux will overwrite an existing target no matter what. So if we want to normalize this behavior, we should probably check for an existing target and unlink it if it exists.

`fs.mkdir` will throw `Error: EEXIST: file already exists` on all platforms if you try and create a dir that already existsw.

**Comment by mekwall** [comment]
The current polyfill will get stuck on Windows at renaming a directory when the target already exists. I think that we should make sure it works the same on both platform (i.e. remove the target directory if it exists).

I tested using the following code: https://gist.github.com/mekwall/b6c44c6855aa9adbd97af1bba9386e50

**Comment by mekwall** [comment]
The last commit normalizes the behavior. This is what I get with the test code from my gist linked above.

Windows:
```
FS
Trying to rename dir
Failed!
GFS
Trying to rename dir
Success!
```

Linux:
```
FS
Trying to rename dir
Success!
GFS
Trying to rename dir
Success!
```

**Comment by vladimiry** [comment]
I've tried this implementation as well as my own atomic writing dependencies free implementation (similar ideas, queue
 + backoff retry) - it doesn't help on my case (I don't really need to handle dirs renaming, just files), EPERM rename error is still happening.

**Comment by vladimiry** [comment]
I was playing around the EPERM issue during the day and it's now clear that the issue's cause is indeed in the resource locking as was written above by @mekwall. This is the way I've got it proven for myself:
- I open the file for reading and close it (release the locking) only after the 20 seconds:
```javascript
        const fd = await promisify(fs.open)(this.file, "r+");

        setTimeout(async () => {
            await promisify(fs.close)(fd);    
        }, 20 * 1000);
```
- Then in another workflow I run `fs.rename` with the target set to the same opened file applying backoff retry for 60 seconds max and I get file renaming successfully completed after the 20 seconds - exactly after releasing the file lock doing `await promisify(fs.close)(fd);`.

So backoff retry scenario should be really helpful, following the @mekwall 's idea. Antiviruses, Dropbox syncing and such kind of software apparently open files for a long time and very frequently, that's why there are so many EPERM issues posted on Github (which usually go to the close state without the actual resolution). So if there is no `retry` scenario, but just a single renaming attempt (or other actions that might cause EPERM, not only the renaming), then we get EPERM on Windows, that's it.

**Comment by vladimiry** [comment]
This is the test case I mentioned above https://github.com/vladimiry/fs-no-eperm-anymore/tree/master/test/locking - locked `fs.rename` call.

**Comment by JoshuaKGoldberg** [comment]
@isaacs do you think you could please take a look?

**Comment by christiango** [comment]
What's the latest on this PR? It would be great to get this in to fix the issues we are seeing with the latest versions of Jest on Windows

**Comment by dogboydog** [comment]
I encounter the EPERM issues very frequently still as do my colleagues.  A colleague recently got a new workstation with windows 10, and could not do npm install with any version of npm he tried.  He got the closest with npm 4.2.0.  We tried blowing out the graceful-fs  dependency used by npm in his installation with the files from this pull request, which fixed his issues. I hope this fix becomes a part of npm soon. Thank you for fixing it

**Comment by RyanCavanaugh** [comment]
Thanks @mekwall for writing this patch! I work on TypeScript, Visual Studio, and Visual Studio Code. The underlying problem described here is coming up *a lot* for VS developers because the TypeScript language service puts a file watcher on the project folder (including when the user is just writing JavaScript). We typically see errors like this:
```
npm ERR!   stack: 'Error: EPERM: operation not permitted, rename \'C:\\foo\\node_modules\\pug-error\\package.json.731450482\' -> \'C:\\foo\\node_modules\\pug-error\\package.json\''
```

As the PR describes, this is definitely related to file watchers - in my testing, if we disable the file watching from the TSLS side, the error stops happening.

I've done some testing and **this PR definitely fixes the issue**. The particular scenario we're trying to make work has gone from failing 1 in 4 times to succeeding the last 35 times in a row.

We really need this to be merged - it's impacting a lot of developers and there's nothing that can be done on our end. Developers running `npm install` while VS or VS Code are open with TS or JS files in the editor end up seeing this error all the time and it's going to result in a lot of support noise on the NPM and VS side.

**Comment by rjgotten** [comment]
@mekwall 

Thank you, thank you, thank you, thank you, thank yo-----

It may take longer to run a big `npm install` pass, but unlike any other solution including all the stuff the NPM team has been throwing up against the wall in a bid to see what sticks, this actually **WORKS**.

**Comment by jonkerfan** [comment]
NPM team really needs to merge this PR ASAP, could any guys in their team give a response?

**Comment by isaacs** [comment]
Hi.  It seems like there's a semantic change here.  If the target directory is not empty, then it should fail with `ENOTEMPTY`.  But, the `fs.rmdir` call (which should error in that case) is ignoring the error.

Also, I'm not sure it's a good idea to spin on CPU for potentially 60 seconds in this case.  Is it possible to make this time shorter, or allow it to be configured?

**Comment by jgoz** [comment]
@mekwall @isaacs Addressed feedback in #131.

**Comment by mekwall** [comment]
Closing this in favor of https://github.com/isaacs/node-graceful-fs/pull/131.

**Review by manjaneqx — COMMENTED** [review]
Good

**Review by manjaneqx — COMMENTED** [review]
Y

**Review by manjaneqx — COMMENTED** [review]
Y

**Pull request closed without merging**